cql3/statements/select_statement.cc

```diff
@@ -1627,15 +1629,39 @@ bool select_statement::is_reversed(const schema& schema) const {
     return is_reversed_;
 }
 
+/// True iff restrictions require ALLOW FILTERING despite there being no coordinator-side filtering.
+static bool needs_allow_filtering_anyway(
+        const restrictions::statement_restrictions& restrictions,
+        db::tri_mode_restriction_t::mode strict_allow_filtering) {
+    using flag_t = db::tri_mode_restriction_t::mode;
+    if (strict_allow_filtering == flag_t::FALSE) {
+        return false;
+    }
+    const auto& ck_restrictions = *restrictions.get_clustering_columns_restrictions();
+    const auto& pk_restrictions = *restrictions.get_partition_key_restrictions();
+    // Even if no filtering happens on the coordinator, we still warn about poor performance when partition
+    // slice is defined but in potentially unlimited number of partitions (see #7608).
+    if ((pk_restrictions.empty() || has_token(pk_restrictions.expression)) // Potentially unlimited partitions.
+        && !ck_restrictions.empty() // Slice defined.
+        && !restrictions.uses_secondary_indexing()) { // Base-table is used. (Index-table use always limits partitions.)
+        if (strict_allow_filtering == flag_t::WARN) {
+            mylog.warn("This query should use ALLOW FILTERING and will be rejected in future versions: {}",
```

Comment:
Note: this warning only appears in the log, not as a warning in the response for the query. Since I don't think this flag is necessary at all, I don't mind, but just wanted to note this.

Reply:
Is there a way to include warnings in a query response?

Reply:
Yes, there is. You can see an example in `create_table_statement.cc`. The statement execution builds a `messages::result_message` and you add a warning to it with the `messages::result_message::add_warning()` method. You should be able to see this warning message in cqlsh in addition to the normal result of your query. I never checked what happens if you do the query from a program.

Reply:
Hm, I wonder if that's what @avikivity meant.

Reply:
Yes. We should adjust the drivers so they default to logging those warnings.

Reply:
(@haaawk )

Reply:
> The statement execution builds a `messages::result_messag`

What about statement preparation?  This check happens at preparation time.

Reply:
Good question. I don't know. I'm guessing that if the CQL protocol has a way to return a warning message, it can be used for either prepare or execute requests, but I really don't know the details - nor do I know of an example where we did this.

Reply:
> I'm guessing that if the CQL protocol has a way to return a warning message, it can be used for either prepare or execute requests

That much is clear.  I can even hack warnings into the prepared_statement object and ram them through the query_processor callstack to ensure they are returned in the result_message for the prepare operation.  But what should happen when that prepared_statement is executed?  Redundantly repeat the same warning?  What if the flag is live-updated to a different state?

Reply:
I don't know. I don't think it makes sense to repeat the same warning twice - if it's generated as part of the prepare, you don't need again as part of every execute. 
BUT, please check what the result looks like in cqlsh (we want to see the warning message, and only see it once) and also in some driver (e.g., cql-pytest using the Python driver) that whatever you do seems reasonable.